Development

Kazuto-neves opened the pull request.

Kazuto-neves committed:
Project Init

Kazuto-neves committed:
correcao

Kazuto-neves committed:
Att ReadMe.md

Kazuto-neves committed:
att

- Remove comment and regions

Natan-Barros commented on JournalEntry.Api/Controllers/JournalEntryController.cs:
```diff
@@ -0,0 +1,76 @@
+        private readonly IJournalEntriesRepository repository;
+
+        private readonly ILogger<JournalEntryController> logger;
+        public JournalEntryController(IJournalEntriesRepository repository, ILogger<JournalEntryController> logger)
```
```suggestion

        public JournalEntryController(IJournalEntriesRepository repository, ILogger<JournalEntryController> logger)
```

Natan-Barros commented on JournalEntry.Api/Controllers/JournalEntryController.cs:
```diff
@@ -0,0 +1,76 @@
+        public JournalEntryController(IJournalEntriesRepository repository, ILogger<JournalEntryController> logger)
+        {
+            this.repository = repository;
+
```
```suggestion
```

Natan-Barros commented on JournalEntry.Api/Controllers/JournalEntryController.cs:
```diff
@@ -0,0 +1,76 @@
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> CreateJournalEntryAsync(EntryDto entryDto)
+        {
+            if (!entryDto.Validate()) return BadRequest();
+
+            await repository.CreateJournalEntryAsync(entryDto.MapToEntry());
+
+            return Accepted();
```
No contrato da spec o endpoint de create seria uma lista de journalEntry e não somente uma. Na parte de Contract https://app.clickup.com/3148423/v/dc/302m7-50422/302m7-1503363

Natan-Barros replied:
Também deve seguir as roles da spec -> A soma de todas as entradas deve ser 0

Natan-Barros commented on JournalEntry.Domain/Enuns/TypeOperationJournalEntry.cs:
```diff
@@ -0,0 +1,19 @@
+    [Flags]
+    public enum TypeOperationJournalEntry
+    {
+        DIVIDENDS = 1,
+        EXPENSES = 2,
+        ASSETS = 4,
+        LIABILITIES = 8,
+        OWNERS_EQUITY = 16,
+        REVENUE = 32
```
Nomenclatura dos enum é em pascal case -> https://learn.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/4x252001(v=vs.71)

Natan-Barros commented on JournalEntry.Domain/Interfaces/IJournalEntriesRepository.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace JournalEntry.Domain.Interfaces
+{
+    public interface IJournalEntriesRepository
+    {
+        Task<IEnumerable<Entry>> GetJournalEntriesAsync();
+        Task<Entry> GetJournalEntryAsync(Guid id);
+        Task CreateJournalEntryAsync(Entry jEntry);
+        Task UpdateJournalEntryAsync(Entry jEntry);
+        Task DeleteJournalEntryAsync(Guid id);
+    }
```
booa, utilizando o Repository Pattern.

Natan-Barros commented on JournalEntry.PropertyTests/JournalEntryPropertyTest.cs:
```diff
@@ -0,0 +1,38 @@
+        private readonly Mock<IJournalEntriesRepository> repositoryStub = new Mock<IJournalEntriesRepository>();
+        private readonly Mock<ILogger<JournalEntryController>> loggerStub = new();
+
+        /*[Property(Arbitrary = new[] {typeof(Types)})]
+        public Property CreateJournalEntriesAsync_WithItemToCreate_ReturnsCreatedJournalEntry()
+        {
+
+        }*/
```
```suggestion
```

Natan-Barros commented on JournalEntry.PropertyTests/JournalEntryPropertyTest.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public class JournalEntryPropertyTest
+    {
+        private readonly Mock<IJournalEntriesRepository> repositoryStub = new Mock<IJournalEntriesRepository>();
+        private readonly Mock<ILogger<JournalEntryController>> loggerStub = new();
+
+        /*[Property(Arbitrary = new[] {typeof(Types)})]
+        public Property CreateJournalEntriesAsync_WithItemToCreate_ReturnsCreatedJournalEntry()
+        {
+
+        }*/
+
+        [Property]
+        public Property Multiply_With_2_Is_The_Same_As_Adding_The_Same_Number(int x)
+        {
+            return (x * 2 == Add(x, x)).ToProperty().Collect(x);
+        }
+
+        private int Add(int x1, int x2) => x1 + x2;
+
+        [Property(Arbitrary = new[] { typeof(Types) })]
+        public Property contaDataTest(DateTime x)
+        {
+            return (x.Month - x.Month == dt(x, x)).ToProperty().Collect(x);
+        }
+
+        private int dt(DateTime x1, DateTime x2) => x1.Month - x2.Month;
+    }
```
A ideia seria testar a journal entry, tenta fazer esse teste utilizando a lib que você estudou.

Natan-Barros commented on JournalEntry.PropertyTests/Services/Types.cs:
```diff
@@ -0,0 +1,42 @@
+        public static Arbitrary<Decimal> DecimalPropTest() => Arb.Default.Decimal().Generator
+            .Where(dec => dec >= 0 && dec < 1001).ToArbitrary();
+
+        /*public static Arbitrary<Object> OperationPropTest() => Arb.Default.Object<OperationJournalEntry>().Generator
+            .Where(oprJE => oprJE == OperationJournalEntry.DEBIT || oprJE == OperationJournalEntry.CREDIT).ToArbitrary();
+
+        public static Arbitrary<EntryDto> EntryDtoPropTest() =>
+            IdPropTest().Generator
+            .Select(
+                    _idPropTest => new EntryDto(
+                    _idPropTest,
+                    Arb.Default.DateTime().Generator.Where(dt => dt > DateTime.UtcNow && dt <= DateTime.UtcNow.AddMonths(3)),
+                    Arb.Default.DateTimeOffset().Generator.Where(dt => dt >= DateTimeOffset.UtcNow),
+                    Arb.Default.Decimal().Generator.Where(dec => dec >= 0 && dec < 1001),
+                    Arb.Default.
+                    //Arb.Default.String().Generator
+                    )
+                ).ToArbitrary();*/
```
```suggestion
```

Natan-Barros commented on JournalEntry.UnitTests/Services/Genereitor.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace JournalEntry.UnitTests.Services
+{
+    public class Genereitor
```
```suggestion
    public class Generator
```

Natan-Barros commented on JournalEntry.Domain/Enuns/OperationJournalEntry.cs:
```diff
@@ -0,0 +1,15 @@
+    [Flags]
+    public enum OperationJournalEntry
+    {
+        CREDIT = 1,
+        DEBIT = 2
```
Aqui também seria pascal case.

Natan-Barros commented on JournalEntry.Api/Controllers/JournalEntryController.cs:
```diff
@@ -0,0 +1,76 @@
+
+        [HttpPut("{id}")]
+        public async Task<IActionResult> UpdateJournalEntryAsync(Guid id, EntryDto entryDto)
+        {
+            var existingJEntry = await repository.GetJournalEntryAsync(id);
+
+            if (existingJEntry is null) return NotFound();
+
+            if (!entryDto.Validate()) return BadRequest();
+
+            await repository.UpdateJournalEntryAsync(entryDto.MapToEntry(id));
+
+            return Accepted();
```
Não estou vendo também as regras especificadas para esse endpoint.

Natan-Barros reviewed (commented):
Fiz uma review superficial, vou liberar esses comentários. Qualquer ponto pode discutir comigo, após eles serem resolvidos vou revisar novamente. Inclusive vou depois analisar a cobertura de teste que ainda não vi.

Kazuto-neves committed:
Update JournalEntry.Api/Controllers/JournalEntryController.cs

Co-authored-by: Natan de Mesquita Barros <47890888+Natan-Barros@users.noreply.github.com>

Kazuto-neves committed:
Update JournalEntry.Api/Controllers/JournalEntryController.cs

Co-authored-by: Natan de Mesquita Barros <47890888+Natan-Barros@users.noreply.github.com>

Kazuto-neves committed:
Update JournalEntry.UnitTests/Services/Genereitor.cs

Co-authored-by: Natan de Mesquita Barros <47890888+Natan-Barros@users.noreply.github.com>

Kazuto-neves committed:
att do journalEntry

- Tentativa de fazer a inseção de N entry de uma vez falha

Kazuto-neves committed:
att journalEntry

. metodos de validador de amount inplementado
. Criando uma lista de journalEntry
. Obs:tem alguns bugs como createDate zerado
. Com certeza os testes estão quebrados

Kazuto-neves committed:
att journalEntry

. Correção na criação
    . Id Zerado
    . DateTimeOfSet Zerado

. Correção do testes Unitarios
    .Create
    .Update

Natan-Barros commented on JournalEntry.Api/Controllers/JournalEntryController.cs:
```diff
@@ -0,0 +1,88 @@
+
+            if (!entriesDto.Validate()) return BadRequest();
+
+            if (!EntryValidatorAddAmount.Validador(entries, entriesDto.MapToEntry())) return BadRequest();
```
```suggestion
            if (!EntriesValidator.ShouldCreateEntries(entries, entriesDto.MapToEntry())) return BadRequest();
```

Kazuto-neves committed:
att

.Retirando alguns warnings

Kazuto-neves committed:
att project

.tratamento dos alertas
.criação de mais testes unitarios

Kazuto-neves committed:
att project

.Criado mais casos de teste Unitario(as falhas esperadas)

Kazuto-neves committed:
att

.deixando o codigo mais generico

Kazuto-neves committed:
att

.abstração do codigo
.adição de todas as Exception no ReturnException

Kazuto-neves committed:
att

. tentativa de fazer callback com o meu conhecimento em callback em Js

Kazuto-neves committed:
att

Kazuto-neves committed:
att

.Tentativa de fazer pelomenos um teste de propriedade funcionar
